File: app/src/main/java/omg/jd/tvmazeapiclient/mvp/PresenterLoader.java
package omg.jd.tvmazeapiclient.mvp;

import android.content.Context;
import android.support.v4.content.Loader;

public class PresenterLoader<T extends BasePresenter> extends Loader<T> {

    private final PresenterFactory<T> factory;
    private T presenter;

    public PresenterLoader(Context context, PresenterFactory<T> factory) {
        super(context);
        this.factory = factory;
    }

    @Override
    protected void onStartLoading() {
        if (presenter != null) {
            deliverResult(presenter);
            return;
        }

        forceLoad();
    }

    @Override
    protected void onForceLoad() {
        // Create the Presenter using the Factory
        presenter = factory.create();

        // Deliver the result
        deliverResult(presenter);
    }

    @Override
    protected void onReset() {
        presenter.onDestroyed();
        presenter = null;
    }
}

File: app/src/main/java/omg/jd/tvmazeapiclient/mvp/PresenterLoader.kt
package omg.jd.tvmazeapiclient.mvp

import android.content.Context
import android.support.v4.content.Loader

class PresenterLoader<T : BasePresenter<*>>(context: Context, private val factory: PresenterFactory<T>) : Loader<T>(context) {
    private var presenter: T? = null

    override fun onStartLoading() {
        if (presenter != null) {
            deliverResult(presenter)
            return
        }

        forceLoad()
    }

    override fun onForceLoad() {
        // Create the Presenter using the Factory
        presenter = factory.create()

        // Deliver the result
        deliverResult(presenter)
    }

    override fun onReset() {
        presenter?.onDestroyed()
        presenter = null
    }
}